pinc/ProofProject.inc

```diff
@@ -0,0 +1,598 @@
+    }
+
+    private function get_page_state_array()
+    {
+        return [
+            "pagestate" => $this->page_state,
+            "saved" => ($this->page_state == $this->round->page_temp_state),
+        ];
+    }
+
+    private function _return_to_round(): void
+    {
+        global $pguser;
+
+        $setters = join(", ", [
+            set_col_str("state", $this->round->page_avail_state),
+            set_col_num($this->round->time_column_name, time()),
+        ]);
+        _Page_UPDATE($this->projectid, $this->page_name, $setters);
+        _log_page_event($this->projectid, $this->page_name, 'returnToRound', $pguser, $this->round);
+        _project_adjust_n_available_pages($this->projectid, +1);
+    }
+}
+
+class ProofProject
+{
+    use ProofPage;
+
+    public function __construct(object $project, array $query_params = [])
+    {
+        global $pguser, $PROJECT_STATES_IN_ORDER;
+
+        $expected_state = $query_params["projstate"] ?? null;
+        if (null === $expected_state) {
+            throw new ProjectNoStateGivenException("No expected state found in request.");
+        }
+        if (!in_array($expected_state, $PROJECT_STATES_IN_ORDER)) {
+            throw new ProjectInvalidStateException(sprintf("%s is not a valid project state", $expected_state));
+        }
+        if ($expected_state != $project->state) {
+            $err = sprintf(
+                _('Warning: Project "%1$s" is not in state "%2$s"; it is now in state "%3$s".'),
+                $project->nameofwork,
+                project_states_text($expected_state),
+                project_states_text($project->state)
+            );
+            throw new ProjectNotInRequiredStateException($err);
+        }
+
+        $project_round = $project->get_project_available_round();
+        $project_round->validate_user_can_access($pguser);
+
+        $this->project = $project;
+        $this->projectid = $project->projectid;
+        $this->round = $project_round;
+    }
+
+    public function checkout_pages(int $requested_pages): array
+    {
+        global $pguser;
+        if ($requested_pages < 1) {
+            throw new InvalidValue("requested pages must be at least 1");
+        }
+        if ($requested_pages > MAX_REQUESTED_PAGES) {
+            throw new ProjectTooManyPagesRequested(sprintf("requested pages must not be more than %d", MAX_REQUESTED_PAGES));
+        }
+        // get some pages. If not enough available return a reason
+        $reason = "";
+        $status_code = 0;
+        $this->validate_no_outpages();
+        $out_pages = 0;
+        validate_user_can_get_pages_in_project($pguser, $this->project, $this->round);
+        while ($out_pages < $requested_pages) {
+            try {
+                $this->checkout_available_proof_page();
+                $out_pages += 1;
+            } catch (ProjectNoMorePagesException | BeginnersQuotaReachedException | BeginnersProjectQuotaReachedException $exception) {
```

Comment:
This is unsatisfactory because future changes to why no more pages can be checked out could break this.